### PR TITLE
GPII-3346 - Enable authentication for CouchDB requests

### PR DESCRIPTION
**Pull request by amatas**
This PR has more explanation about tests that I did than code pushed, as you can see.

Enable authentication for the port 5984 its easy, just one variable must set in the configuration:
http://docs.couchdb.org/en/latest/config/auth.html#chttpd/require_valid_user

The problem is that the liveness and healthcheck proves doesn't work. As they use `httpGet` to some urls they need also authenticate to check if the service is ok.
Possible solutions
1. ~~Authenticate the HTTP test request. [The proves](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes) only support a few parameters which don't allow authentication. But one of the parameters `httpHeaders` [can be used to provide authentication credentials](https://stackoverflow.com/questions/33484942/how-to-use-basic-authentication-in-a-http-liveness-probe-in-kubernetes). The problem is how to build the base64 hash using Kubernetes secrets variables interpolation. As the hash must be created at the creation of the cluster and in the statefulset context. We can create the headers and the hash either in TF code or Rake code but this not seem very scalable, overall if in the future we use key rotation.~~ . It's not easy to implement, requires additional variables for storing the base64 hash. And all the tests that I did didn't work as expected. Despite the header was set, the httpGet wasn't able to authenticate.

1. ~~Change the "code return" that validates the test. It seems that the `httpGet` doesn't support that.~~ It is not possible.
1. ~~Use the prove to `tcpSocket` instead. The port could be open, but the CouchDB can not be ready for serving data.~~ it doesn't cover all the cases (see comments)
1. ~~Change the test port to 5986 instead. The 5986 and the 5984 ports are handled with different options in the configuration so the 5986 would allow unauthenticated checks but is not exposed outside the statefulset with fw rules. This is what is implemented in this PR as it seems the simpler solution.~~ But for some reason the CouchDB dies, and complains about the lack of the _user database:
    ```

    [error] 2018-09-28T16:22:11.702832Z couchdb@couchdb-couchdb-0.couchdb-couchdb.gpii.svc.cluster.local emulator -------- Error in process <0.632.0> on node 'couchdb@couchdb-couchdb-0.couchdb-couchdb.gpii.svc.cluster.local' with exit value:                                              
    {database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,395}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,370}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,399}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,86}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,38}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,187}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,134}]}]}
    ```
    The only way to bypass this error is to create the `_user` database, but this requires some interaction with the CouchDB container, so I guess that the proper place to handle this is in the sidecar container.

    This didn't work, the `_up` endpoint is not available at the admin port.

1. The last solution that worked was to use a `exec` check, which uses a curl command that runs inside the couchdb container, it allows the use of the environment variables so it is not necessary the use of additional variables.

### CODE REVIEW

**Comment by mrtyler** [comment]
Good comments from Stepan. Adding a few notes:

I'm skeptical of solution 1 because it doesn't seem like something upstream could use. A variant that might work is to store the required credentials in a Kubernetes Secret and reference the values from the Probe configuration (if that's possible).

For solution 3, I think the existing liveness probe is causing https://issues.gpii.net/browse/GPII-3397 and needs to be changed regardless. A tcpSocket probe for liveness might be better. (I agree with Stepan that we should stick with http[s] for the readiness check).

Another option might be an ExecAction probe: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes

I'm ok with solution 4, though I agree with Stepan that we don't want to expose any ports unnecessarily. And, I think we want a solution that is usable by anyone so that upstream can accept our changes.

File: gcp/modules/couchdb/values.yaml [hunk 1]
```diff
@@ -21,3 +21,9 @@ affinity:
               values:
               - couchdb
         topologyKey: "kubernetes.io/hostname"
+
+couchdbConfig:
+  chttpd:
+    bind_address: any
+    # http://docs.couchdb.org/en/stable/config/auth.html#authentication-configuration
+    require_valid_user: true
```
Comment (natarajaya) [inline gcp/modules/couchdb/values.yaml]: Please configure your editor to add trailing newlines automatically.

**Comment by natarajaya** [comment]
Using node-local port instead of exposed clustered-port for unauthenticated lifecycle probes makes sense!
Regarding solution to GPII-3397, switching liveness probe to `TCPSocketAction` is probably most straight-forward thing to do (which also looks good enough to propose to upstream).

**Comment by amatas** [comment]
The `_up` endpoint is not available at the admin port, so the solution number 4 is not possible. I've explored a 5 solution, which is the use of ExecAction probes. I use the curl internal command with the environment variables provided so it does not need additional secrets as solution 1. The implementation is in https://github.com/gpii-ops/gpii-infra/pull/150/commits/f362f7ce4f822207f8a63f9d1ae94f7db9027e12 commit

**Comment by mrtyler** [comment]
LGTM

I would probably change the livenessProbe to tcpSocket to fix GPII-3397 while I was in here, but it's ok if you want to handle that separately.

**Comment by amatas** [comment]
@mrtyler Let me test GPII-3397 separately, I'll open a new PR for it.

**Comment by amatas** [comment]
@mrtyler @natarajaya @stepanstipl please, check the latest changes and the if conditions in the templates. If everything looks good I'll merge this tomorrow morning.

File: shared/charts/couchdb/templates/statefulset.yaml [hunk 2]
```diff
@@ -61,13 +61,29 @@ spec:
             - name: ERL_FLAGS
               value: "{{ range $k, $v := .Values.erlangFlags }} -{{ $k }} {{ $v }} {{ end }}"
           livenessProbe:
+{{- if .Values.couchdbConfig.chttpd.require_valid_user }}
```
Comment (stepanstipl) [inline shared/charts/couchdb/templates/statefulset.yaml]: Why do we have the split behavior - is there any disadvantage to changing this to exec every time?
 
What I'm thinking is that while the result **_should_** be the same, why introduce the possibility of different behavior?

File: shared/charts/couchdb/templates/statefulset.yaml [hunk 3]
```diff
@@ -61,13 +61,29 @@ spec:
...
+            exec:
+              command:
+                - sh
+                - -c
+                - curl -G --silent --fail -u ${COUCHDB_USER}:${COUCHDB_PASSWORD} http://localhost:${COUCHDB_SVC_COUCHDB_SERVICE_PORT}
```
Comment (stepanstipl) [inline shared/charts/couchdb/templates/statefulset.yaml]: Where is the `COUCHDB_SVC_COUCHDB_SERVICE_PORT` variable coming from? It seems the port it's hardcoded at some places in the chart and not in others (like the httpGet check right below). I would vote for consistency across the chart.

**Comment by mrtyler** [comment]
LGTM

**Comment by natarajaya** [comment]
LGTM

File: shared/charts/couchdb/templates/statefulset.yaml (hunk 2 above)
Comment (amatas) [reply shared/charts/couchdb/templates/statefulset.yaml]: I was afraid of having errors by using credentials when they are not needed. That's why I prefer to use the old check.

**Pull request merged**